Release process: automatically close milestone when starting the release

tarunvijwani opened the pull request:
This PR adds more functionality to the release PR automation, it closes the existing milestone of the release and creates the next milestone. 

## Changes
- Added the code to close existing milestones and create the next milestone in `lib/automations/release/branch-create-handler.js`


# Testing
To test this, you will need to:
- **Fork** the `woocommerce-gutenberg-products-block` and then unfork it (Go to this url: https://support.github.com/contact?tags=rr-forks and type `unfork` into the subject line. Follow the steps from the virtual assistant and wait 5-10 mins for the request to be actioned).
- Enable issues on your new unforked repository (go to `Settings > General > Features > Issues`)
- Activate GH actions
- Give read and write permissions to the workflows: Settings > Actions > General > Workflow permissions
- Allow GH actions to create PRs: Settings > Actions > General
- Create a milestone (Example: 11.1.0)
- Clone your repository locally.
- Create a PR on your repository, in this PR, you should change the following file: https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/trunk/.github/workflows/release-pull-request.yml#L27
- Change the highlighted line to read: `uses: woocommerce/automations@add/close-milestone`
- Merge the PR
- Create a new branch `release/11.1.0` (same as your milestone number)
- Go to actions and confirm the `Release Pull Request Automation ` task is completed successfully. 
- Go to milestones and confirm existing milestone is closed, and the next milestone is created successfully.

ralucaStan commented:
@tarunvijwani this PR includes some changes to package-lock.json that are not relevant for this PR. There is also the [dist/index.js](https://github.com/woocommerce/automations/pull/54/files#diff-3d2b59189eeedc2d428ddd632e97658fe310f587f7cb63b01f9b98ffc11c0197) file that shouldn't have been included here.

tarunvijwani commented:
> @tarunvijwani this PR includes some changes to package-lock.json that are not relevant for this PR. There is also the [dist/index.js](https://github.com/woocommerce/automations/pull/54/files#diff-3d2b59189eeedc2d428ddd632e97658fe310f587f7cb63b01f9b98ffc11c0197) file that shouldn't have been included here.

@ralucaStan We need build files in dist in order to test and run the PR. Without them, it won't work. I have added the dist folder changes again to the PR.

gigitux commented:
Hey @tarunvijwani!
I tried to run the automation, but I have this error ([job](https://github.com/gigitux/woocommerce-blocks/actions/runs/3393829572/jobs/5641619904)):
<img width="1074" alt="image" src="https://user-images.githubusercontent.com/4463174/199975935-f433e91f-2335-458b-9e5e-2bee5a8407cd.png">

Do you know how I can fix it?

> We need build files in dist in order to test and run the PR. Without them, it won't work. I have added the dist folder changes again to the PR.

Yep, building files in the `dist` folder is necessary.

Also, before proceeding with the release, we should fix the unit test suite. I think [this PR](https://github.com/woocommerce/automations/pull/51) introduces some failures in the unit tests. cc @opr

Commit:
rebase trunk

tarunvijwani commented:
> Hey @tarunvijwani! I tried to run the automation, but I have this error ([job](https://github.com/gigitux/woocommerce-blocks/actions/runs/3393829572/jobs/5641619904)): <img alt="image" width="1074" src="https://user-images.githubusercontent.com/4463174/199975935-f433e91f-2335-458b-9e5e-2bee5a8407cd.png">
> 
> Do you know how I can fix it?
> 
> > We need build files in dist in order to test and run the PR. Without them, it won't work. I have added the dist folder changes again to the PR.
> 
> Yep, building files in the `dist` folder is necessary.

I tried rebasing the trunk and force-commit my changes, but it didn't help. It is related to readme.txt. @tarhi-saad @opr Could you please test this PR in your repo and check? 


> Also, before proceeding with the release, we should fix the unit test suite. I think [this PR](https://github.com/woocommerce/automations/pull/51) introduces some failures in the unit tests. cc @opr

Some of the changes of #51 were overwritten by #53 . We need to check both PRs.

opr commented on lib/automations/release/branch-create-handler.js:
```diff
@@ -328,6 +328,60 @@ const branchHandler = async ( context, octokit, config ) => {
+	);
+	const milestoneNumber = milestone.number;
+	const milestoneTitle = milestone.title
+		.replace( '"', '' )
+		.replace( '[', '' )
+		.replace( ']', '' );
```
Why do we need to do this?

You can also change to

```js
milestone.title.replace( /["\[\]]/g, '' );
```

if you prefer, the way you've done it might be more readable though.

tarunvijwani replied:
Thank you, I added the regex and updated the code. 👍

opr commented on lib/automations/release/branch-create-handler.js:
```diff
@@ -328,6 +328,60 @@ const branchHandler = async ( context, octokit, config ) => {
+		.replace( '[', '' )
+		.replace( ']', '' );
+	// Close existing milestone
+	const milestoneUpdate = await octokit.issues.updateMilestone( {
```
What do you think about also updating the `due_on` date during this update, we set the milestone's due date to the ship date when releasing, so this could save us a step.

tarunvijwani replied:
@opr That's a good point. We can indeed do that. However, there may be times when we don't ship the release on the day we create the release branch. It may take a day or two to complete testing and address any blockers. In those cases, the due date would be incorrect.

opr reviewed (approved):
Hey Tarun, works great for me, thanks for working on it!

I left a couple of comments, but I don't think they're blocking, feel free to work on them or merge and work on them later.

Commit:
Add regex to remove quotes and brackets from milestone title

Using regex instead of string replace to remove quotes and brackets from milestone title
